crates/qpy/src/circuit_reader.rs

```diff
@@ -0,0 +1,995 @@
+    for packed_register in non_standalone_registers {
+        match packed_register.register_type {
+            register_types::QREG => {
+                let bits: Vec<ShareableQubit> = packed_register
+                    .bit_indices
+                    .iter()
+                    .filter_map(|&index| {
+                        if index >= 0 {
+                            Some(final_qubit_list[index as usize].clone())
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                let qreg = QuantumRegister::new_alias(Some(packed_register.name.clone()), bits);
+                qregs.push(qreg);
+            }
+            register_types::CREG => {
+                let bits: Vec<ShareableClbit> = packed_register
+                    .bit_indices
+                    .iter()
+                    .filter_map(|&index| {
+                        if index >= 0 {
+                            Some(final_clbit_list[index as usize].clone())
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                let creg = ClassicalRegister::new_alias(Some(packed_register.name.clone()), bits);
+                cregs.push(creg);
+            }
+            _ => {
+                return Err(PyValueError::new_err(format!(
+                    "Unrecognized register type for {:?}",
+                    packed_register.name
+                )));
+            }
+        }
+    }
+
+    // now add the bits to the ciruit, and then add the registers
+    for qubit in final_qubit_list {
+        qpy_data.circuit_data.add_qubit(qubit, true)?;
+    }
+    for clbit in final_clbit_list {
+        qpy_data.circuit_data.add_clbit(clbit, true)?;
+    }
+
+    for qreg in qregs {
+        qpy_data.circuit_data.add_qreg(qreg, true)?;
+    }
+
+    for creg in cregs {
+        qpy_data.circuit_data.add_creg(creg, true)?;
+    }
+
+    Ok(())
+}
+
+pub fn unpack_circuit(
+    py: Python,
+    packed_circuit: &QPYFormatV15,
+    version: u32,
+    metadata_deserializer: Option<&Bound<PyAny>>,
+    use_symengine: bool,
+    annotation_factories: &Bound<PyDict>,
+) -> PyResult<Py<PyAny>> {
+    let instruction_capacity = packed_circuit.instructions.len();
+    // create an empty circuit; we'll fill data as we go along
+    let mut circuit_data =
+        CircuitData::with_capacity(0, 0, instruction_capacity, Param::Float(0.0))?;
+
+    let annotation_handler = AnnotationHandler::new(annotation_factories);
+    let mut qpy_data = QPYReadData {
+        circuit_data: &mut circuit_data,
+        version,
+        use_symengine,
+        standalone_vars: HashMap::new(),
+        standalone_stretches: HashMap::new(),
+        vectors: HashMap::new(),
+        annotation_handler,
+    };
```

Comment:
Minor style suggestion:
```suggestion
    let mut qpy_data = QPYReadData {
        circuit_data: &mut circuit_data,
        version,
        use_symengine,
        standalone_vars: HashMap::new(),
        standalone_stretches: HashMap::new(),
        vectors: HashMap::new(),
        annotation_handler: AnnotationHandler::new(annotation_factories)
    };
```